File: app/javascript/lib/share-text.js
import { parseSlug, passageLabel } from "./passage-span.js"
import { wikiTokens } from "./wiki-markup.js"

const LINK_BASE = "https://route.bible"

export function wikiToPlain(text) {
  return wikiTokens(text).map((token) => (
    token.type === "wiki" ? token.label : token.value
  )).join("")
}

export function noteLines(blocks, baseIndent = 0, bullets = false) {
  return (Array.isArray(blocks) ? blocks : [])
    .map((block) => {
      const text = String(block?.text || "").replace(/\s+$/g, "")
      if (!text.trim()) return null
      const indent = baseIndent + (Number(block.indent) || 0)
      const prefix = "  ".repeat(indent)
      const mark = bullets && block.bullet !== false ? "- " : ""
      return `${prefix}${mark}${text}`
    })
    .filter(Boolean)
}

export function formatNoteShare({ label, blocks }) {
  const lines = []
  if (label) lines.push(String(label).trim())
  const body = noteLines(blocks, 0, true)
  if (body.length) {
    if (lines.length) lines.push("")
    lines.push(...body)
  }
  return `${lines.join("\n").trim()}\n`
}

export function formatVerseShare({ label, text, notes, url }) {
  const lines = [ label, wikiToPlain(text || "").trim() ]
  for (const note of Array.isArray(notes) ? notes : []) {
    const body = noteLines(note.blocks, 0, true)
    if (body.length) lines.push("", ...body)
  }
  if (url) lines.push("", url)
  return lines.join("\n").trim() + "\n"
}

export function formatChapterShare({ label, chapterNote, verses, url, bullets = false }) {
  const lines = [ label ]
  const chapterBody = noteLines(chapterNote, 0, bullets)
  if (chapterBody.length) lines.push("", ...chapterBody)

  for (const verse of Array.isArray(verses) ? verses : []) {
    if (verse.heading) lines.push("", verse.heading)
    lines.push("", `${verse.n}. ${wikiToPlain(verse.text || "").trim()}`)
    for (const note of Array.isArray(verse.notes) ? verse.notes : []) {
      const body = noteLines(note.blocks, 0, bullets)
      if (body.length) lines.push(...body)
    }
  }
  if (url) lines.push("", url)
  return lines.join("\n").trim() + "\n"
}

export function escapeHtml(value) {
  return String(value)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
}

export function inlineToHtml(text) {
  return wikiTokens(text).map((token) => {
    if (token.type !== "wiki") {
      return escapeHtml(token.value).replace(/\n/g, "<br>")
    }
    const label = escapeHtml(token.label)
    if (!token.slug) return label
    return `<a href="${escapeHtml(`${LINK_BASE}/${token.slug}`)}">${label}</a>`
  }).join("")
}

export function blocksToHtml(blocks, baseIndent = 0) {
  const items = (Array.isArray(blocks) ? blocks : [])
    .map((block) => {
      const text = String(block?.text || "").replace(/\s+$/g, "")
      if (!text.trim()) return null
      return {
        indent: baseIndent + (Number(block.indent) || 0),
        html: inlineToHtml(text)
      }
    })
    .filter(Boolean)
  if (!items.length) return ""

  let html = ""
  let level = -1
  for (const item of items) {
    if (item.indent > level) {
      html += "<ul>"
      level = item.indent
    } else if (item.indent === level) {
      html += "</li>"
    } else {
      while (level > item.indent) {
        html += "</li></ul>"
        level -= 1
      }
      html += "</li>"
    }
    html += `<li>${item.html}`
  }
  while (level >= 0) {
    html += "</li></ul>"
    level -= 1
  }
  return html
}

export function formatNoteHtml({ label, blocks }) {
  const parts = []
  if (label) parts.push(`<p><strong>${escapeHtml(String(label).trim())}</strong></p>`)
  const list = blocksToHtml(blocks)
  if (list) parts.push(list)
  return parts.join("")
}

export function formatChapterHtml({ label, chapterNote, verses }) {
  const parts = []
  if (label) parts.push(`<p><strong>${escapeHtml(label)}</strong></p>`)
  const chapterList = blocksToHtml(chapterNote)
  if (chapterList) parts.push(chapterList)
  for (const verse of Array.isArray(verses) ? verses : []) {
    if (verse.heading) parts.push(`<p><strong>${escapeHtml(verse.heading)}</strong></p>`)
    parts.push(`<p>${verse.n}. ${escapeHtml(wikiToPlain(verse.text || "").trim())}</p>`)
    for (const note of Array.isArray(verse.notes) ? verse.notes : []) {
      const list = blocksToHtml(note.blocks)
      if (list) parts.push(list)
    }
  }
  return parts.join("")
}

export function formatBookShare({ label, chapters, url }) {
  const parts = [ label ]
  for (const chapter of Array.isArray(chapters) ? chapters : []) {
    parts.push("", formatChapterShare({
      label: chapter.label,
      chapterNote: chapter.chapterNote,
      verses: chapter.verses,
      bullets: true
    }).trimEnd())
  }
  if (url) parts.push("", url)
  return parts.join("\n").trim() + "\n"
}

export function passageUrl(slug) {
  return slug ? `${LINK_BASE}/${slug}` : ""
}

export function verseShareLabel(bookLabel, chapter, start, end) {
  return passageLabel(bookLabel, chapter, start, end)
}

export function notesForVerse(notes, chapterSlug, n) {
  const list = Array.isArray(notes) ? notes : []
  const exact = `${chapterSlug}.${n}`
  return list.filter((note) => {
    const parsed = parseSlug(note.slug)
    if (!parsed || parsed.kind === "chapter") return false
    if (note.slug === exact) return true
    return parsed.kind === "range" && parsed.verseEnd === n
  })
}

File: lib/margin/share_text.rb
# frozen_string_literal: true

module Margin
  module ShareText
    LINK_BASE = "https://route.bible"
    WIKI = /\[\[([^\]|]+)(?:\|([^\]]+))?\]\]/

    module_function

    def wiki_to_plain(text)
      text.to_s.gsub(WIKI) { |match|
        label = Regexp.last_match(2)
        target = Regexp.last_match(1)
        label.presence || Passage.parse(target)&.label || target
      }
    end

    def note_lines(blocks, base_indent: 0, bullets: false)
      Array(blocks).filter_map { |block|
        text = wiki_to_plain(block.is_a?(Hash) ? block["text"] : block[:text]).rstrip
        next if text.blank?

        indent = base_indent + (block.is_a?(Hash) ? block["indent"] : block[:indent]).to_i
        prefix = "  " * indent
        bullets ? "#{prefix}- #{text}" : "#{prefix}#{text}"
      }
    end

    def format_verse(label:, text:, notes: [], url: nil)
      lines = [ label, wiki_to_plain(text).strip ]
      Array(notes).each do |note|
        body = note_lines(note_blocks(note), base_indent: 1)
        next if body.empty?

        lines << ""
        lines.concat(body)
      end
      lines << "" << url if url.present?
      "#{lines.join("\n").strip}\n"
    end

    def format_chapter(label:, verses:, chapter_note: nil, url: nil, bullets: false)
      lines = [ label ]
      chapter_body = note_lines(note_blocks(chapter_note), base_indent: 0, bullets:)
      if chapter_body.any?
        lines << ""
        lines.concat(chapter_body)
      end
      Array(verses).each do |verse|
        heading = verse[:heading] || verse["heading"]
        lines << "" << heading if heading.present?
        n = verse[:n] || verse["n"]
        text = verse[:text] || verse["text"]
        lines << "" << "#{n}. #{wiki_to_plain(text).strip}"
        Array(verse[:notes] || verse["notes"]).each do |note|
          body = note_lines(note_blocks(note), base_indent: 1, bullets:)
          lines.concat(body) if body.any?
        end
      end
      lines << "" << url if url.present?
      "#{lines.join("\n").strip}\n"
    end

    def format_book(label:, chapters:, url: nil, bullets: false)
      parts = [ label ]
      Array(chapters).each do |chapter|
        parts << ""
        parts << format_chapter(
          label: chapter[:label] || chapter["label"],
          verses: chapter[:verses] || chapter["verses"],
          chapter_note: chapter[:chapter_note] || chapter["chapterNote"],
          bullets:
        ).strip
      end
      parts << "" << url if url.present?
      "#{parts.join("\n").strip}\n"
    end

    def document(scope:, book:, chapter: nil, verse_start: nil, verse_end: nil, notes: {}, include_notes: true, include_url: false)
      notes = normalize_notes(notes)
      notes = {} unless include_notes
      case scope.to_s
      when "verse", "range"
        verse_document(book:, chapter:, verse_start:, verse_end:, notes:, include_url:)
      when "chapter"
        chapter_document(book:, chapter:, notes:, include_url:)
      when "book"
        book_document(book:, notes:, include_url:)
      when "bible"
        bible_document(notes:, include_url:)
      else
        nil
      end
    end

    def filename(scope:, book: nil, include_notes: true)
      suffix = include_notes ? "-notes" : ""
      case scope.to_s
      when "bible" then "bible#{suffix}.md"
      when "book"
        slug = (Books.name_for(book) || book).to_s.parameterize
        "#{slug}#{suffix}.md"
      else
        "passage#{suffix}.md"
      end
    end

    def verse_document(book:, chapter:, verse_start:, verse_end:, notes:, include_url:)
      passage = Passage.new(book:, chapter:, verse_start:, verse_end:)
      data = Bsb.chapter_from_pack(book, chapter)
      return unless data

      verses = pack_verses(data, notes, passage.chapter_slug, from: passage.verse_start, to: passage.span_end)
      url = include_url ? "#{LINK_BASE}/#{passage.slug}" : nil
      if passage.kind == "verse"
        row = verses.first
        return unless row

        format_verse(label: passage.label, text: row[:text], notes: row[:notes], url:)
      else
        format_chapter(label: passage.label, verses:, url:)
      end
    end

    def chapter_document(book:, chapter:, notes:, include_url:)
      passage = Passage.new(book:, chapter:)
      data = Bsb.chapter_from_pack(book, chapter)
      return unless data

      format_chapter(
        label: passage.label,
        chapter_note: notes[passage.slug],
        verses: pack_verses(data, notes, passage.slug),
        url: include_url ? "#{LINK_BASE}/#{passage.slug}" : nil
      )
    end

    def book_document(book:, notes:, include_url:)
      code = book.to_s.upcase
      name = Books.name_for(code) || code
      chapters = 1.upto(Books.chapter_count(code)).filter_map { |ch|
        data = Bsb.chapter_from_pack(code, ch)
        next unless data

        slug = "#{code.downcase}.#{ch}"
        {
          label: Passage.new(book: code, chapter: ch).label,
          chapter_note: notes[slug],
          verses: pack_verses(data, notes, slug)
        }
      }
      format_book(
        label: name,
        chapters:,
        url: include_url ? "#{LINK_BASE}/#{code.downcase}.1" : nil,
        bullets: true
      )
    end

    def bible_document(notes:, include_url:)
      parts = [ "Holy Bible" ]
      Books::CODES.each do |code|
        parts << ""
        parts << book_document(book: code, notes:, include_url: false).strip
      end
      parts << "" << LINK_BASE if include_url
      "#{parts.join("\n").strip}\n"
    end

    def pack_verses(data, notes, chapter_slug, from: nil, to: nil)
      Array(data["verses"]).filter_map { |vr|
        n = vr["v"].to_i
        next if from && n < from
        next if to && n > to

        {
          n:,
          heading: vr["heading"].presence,
          text: vr["text"].to_s,
          notes: notes_for_verse(notes, chapter_slug, n)
        }
      }
    end

    def notes_for_verse(notes, chapter_slug, n)
      exact = "#{chapter_slug}.#{n}"
      list = []
      list << notes[exact] if notes[exact]
      notes.each do |slug, blocks|
        next if slug == exact

        parsed = Passage.parse(slug)
        next unless parsed&.range? && parsed.span_end == n && parsed.chapter_slug == chapter_slug

        list << blocks
      end
      list.map { |blocks| { blocks: note_blocks(blocks) } }
    end

    def normalize_notes(notes)
      case notes
      when Hash
        notes.each_with_object({}) { |(slug, value), acc|
          key = slug.to_s
          acc[key] = value.is_a?(Hash) ? (value["blocks"] || value[:blocks] || value) : value
        }
      when Array
        notes.each_with_object({}) { |note, acc|
          slug = note.respond_to?(:slug) ? note.slug : note[:slug] || note["slug"]
          next if slug.blank?

          acc[slug.to_s] = note_blocks(note)
        }
      else
        {}
      end
    end

    def note_blocks(note)
      return [] if note.blank?
      return note if note.is_a?(Array)
      return note["blocks"] || note[:blocks] || [] if note.is_a?(Hash)
      return note.blocks if note.respond_to?(:blocks)

      []
    end
  end
end
